### ISSUE TITLE
set_local_address on request

### ISSUE THREAD

**Issue by johndoe52** [hyperium/hyper#1683]
hello,
as a new hyper client creates a new thread pool etc.
it'd be nice to be able to use set_local_address if you want to make concurrent requests to an host from several addresses that are assigned on the machine and still efficiently use the thread pool..

I basically have to send thousands of requests concurrently to an host with any IP from a range that's assigned on the machine

not sure what's the best way to do this

thanks

**Comment by johndoe52** [comment]
> There is [`HttpConnector::set_local_address`](https://docs.rs/hyper/0.12.12/hyper/client/struct.HttpConnector.html#method.set_local_address) already.
> 
> Or do you mean setting it differently depending on the destination? If so, you could try implementing a custom [`Connect`](https://docs.rs/hyper/0.12.*/hyper/client/connect/index.html) that is tailored for your workflow.

Yes the API exists but requires creating a new client which destroys efficiency.
Setting source address directly when making a request with hyper would be better (not in client builder)

**Comment by seanmonstar** [comment]
> Yes the API exists but requires creating a new client which destroys efficiency. Setting source address directly when making a request with hyper would be better (not in client builder).

Your use case sounds possible to support, but also a bit detailed. That's OK! That's why the `Client` is generic over a connector. You can implement [`Connect`](https://docs.rs/hyper/0.12.*/hyper/client/connect/index.html) on a new type, such that you can inspect the `Destination::hostname()`, and from that, configure a `TcpStream` with a certain local address, and then connect.

Is there a reason this wouldn't work in your situation?

**Comment by johndoe52** [comment]
> > Yes the API exists but requires creating a new client which destroys efficiency. Setting source address directly when making a request with hyper would be better (not in client builder).
> 
> Your use case sounds possible to support, but also a bit detailed. That's OK! That's why the `Client` is generic over a connector. You can implement [`Connect`](https://docs.rs/hyper/0.12.*/hyper/client/connect/index.html) on a new type, such that you can inspect the `Destination::hostname()`, and from that, configure a `TcpStream` with a certain local address, and then connect.
> 
> Is there a reason this wouldn't work in your situation?

I will experiment with this.
I'm looking forward to have this merged directly in Hyper though!

**Comment by johndoe52** [comment]
> My feeling is that this functionality is fairly niche, so I'm going to close the issue for now. If there's lots of need that I don't know about, we can reconsider!

For example, building a web spider over hyper would require this to scale.
